tests/test/existential_types.rs

```diff
@@ -121,6 +121,66 @@ fn dyn_Foo_Bar() {
     }
 }
 
+#[test]
+fn dyn_super_trait_simple() {
+    test! {
+        program {
+            trait Foo<T> {}
+            trait Bar<T> where Self: Foo<T> {}
+
+            struct A {}
+            struct B {}
+        }
+
+        goal {
+            dyn Bar<A>: Bar<A>
+        } yields {
+            "Unique"
+        }
+
+        goal {
+            dyn Bar<A>: Foo<A>
+        } yields {
+            "Unique"
+        }
+
+        goal {
+            dyn Bar<A>: Foo<B>
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            exists<T> {
+                dyn Bar<T>: Foo<B>
+            }
+        } yields {
+            "Unique; substitution [?0 := B], lifetime constraints []"
+        }
+    }
+}
+
+#[test]
+fn dyn_super_trait_cycle() {
+    test! {
+        program {
+            trait Foo<T> where Self: Bar<T> {}
+            trait Bar<T> where Self: Foo<T> {}
+
+            struct A {}
+            struct B {}
+        }
+
+        // It doesn't matter that much what comes out here since such a cycle is
+        // disallowed; mostly this is making sure that we don't crash
```

Comment:
I didn't really investigate, but my guess is because you get a cycle of having to prove `dyn Foo<T>: Bar<T>` to prove `dyn Foo<T>: Foo<T>` to prove `dyn Foo<T>: Bar<T>` and so on... But I don't 100% understand how cycles in the logic are handled, so maybe that should actually be proveable...

Reply:
Oh, yeah, of course, that makes sense. That sort of cycle would be considered unprovable today at least.